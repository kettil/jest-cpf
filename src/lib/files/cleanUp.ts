export const cleanUpByPatterns = (files: string[], patterns: string[], rootPath: string) =>
  patterns
    .map((pattern) => new RegExp(pattern.replace('<rootDir>', rootPath)))
    .reduce((reduceFiles, pattern) => reduceFiles.filter((file) => !pattern.test(file)), files);

export const cleanUpByFiles = (files: string[], testFiles: string[]) => {
  testFiles.forEach((file) => {
    const pos = files.indexOf(file);

    if (pos >= 0) {
      delete files[pos];
    }
  });

  return files.filter((f) => Boolean(f));
};
