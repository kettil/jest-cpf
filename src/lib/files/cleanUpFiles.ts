import { JestConfigType } from '../types';

export const cleanUpByPatterns = (files: string[], patterns: string[], rootPath: string) =>
  patterns
    .map((pattern) => new RegExp(pattern.replace('<rootDir>', rootPath)))
    .reduce((reduceFiles, pattern) => reduceFiles.filter((file) => !pattern.test(file)), files);

export const cleanUpByFiles = (files: string[], testFiles: string[]) => {
  testFiles.forEach((file) => {
    const pos = files.indexOf(file);

    if (pos >= 0) {
      delete files[pos];
    }
  });

  return files.filter((f) => Boolean(f));
};

const cleanUpFiles = (files: string[], testFiles: string[], config: JestConfigType) => {
  // remove all files from ignore patterns
  const filesWithoutPatterns = cleanUpByPatterns(files, config.coveragePathIgnorePatterns, config.rootDir);

  // remove all test files
  return cleanUpByFiles(filesWithoutPatterns, testFiles);
};

export default cleanUpFiles;
