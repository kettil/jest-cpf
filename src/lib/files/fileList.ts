import { join } from 'path';
import glob from 'glob';

const getFileList = (pattern: string, cwd: string, path: string) =>
  new Promise<string[]>((resolve, reject) => {
    glob(pattern, { cwd }, (error, matches) =>
      error ? reject(error) : resolve(matches.map((match) => join(path, match))),
    );
  });

export default getFileList;
