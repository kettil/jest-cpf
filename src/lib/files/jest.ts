import { constants } from 'fs';
import { join, dirname } from 'path';
import { hasFileAccess } from './fileSystems';

export const getJestConfig = async (cwdDirectory: string, filename: string): Promise<Record<string, unknown>> => {
  const file = join(cwdDirectory, filename);

  const isReadable = await hasFileAccess(file, constants.R_OK);

  if (!isReadable) {
    throw new Error(`File ${file} is not readable`);
  }

  return require(file);
};

export const getJestFilename = async (path: string) => {
  let isRoot = false;
  let pathname = path;

  do {
    const filename = join(pathname, 'node_modules/.bin/jest');
    const isExecutable = await hasFileAccess(filename, constants.X_OK);

    if (isExecutable) {
      return filename;
    }

    isRoot = pathname === dirname(pathname);
    pathname = dirname(pathname);
  } while (!isRoot && pathname !== '');

  return 'jest';
};
