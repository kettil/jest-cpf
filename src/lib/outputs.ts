/* eslint-disable no-console */
import { SingleBar } from 'cli-progress';
import { table } from 'table';
import { TestError } from './helpers/errors';
import { padLeft, mapPerLine } from './helpers/string';
import styling from './helpers/styling';
import createTableData from './threshold/createTableData';
import { tableConfig } from './threshold/settings';
import { ThresholdColorType, ThresholdGroupType, BarReturnType, ThresholdType } from './types';

export const outputError = (error: unknown) => {
  if (error instanceof Error) {
    if (error instanceof TestError) {
      console.error(styling('|> An error was thrown during testing!', 'red'));
      console.error(styling('|> ', 'red'));
      console.error(styling('|> Output', 'red'));
      console.error(styling('|> ', 'red'));
      console.error(mapPerLine(error.data, (line) => styling('|> ', 'red') + line));
    } else {
      console.error(mapPerLine(error.stack ?? `Error: ${error.message}`, (line) => styling('|> ', 'red') + line));
    }
  } else {
    console.error(error);
  }

  console.log('');
};

export const outputMetrics = (status: Record<ThresholdColorType, number>, runtime: string) => {
  const types = Object.keys(status) as ThresholdColorType[];
  const maxChars = types.reduce((p, c) => Math.max(p, status[c]), 0).toString().length;

  console.log(styling('Summary', 'underline'));
  console.log(`├─ Success: ${styling(padLeft(status.success.toString(), maxChars, ' '), 'green')}`);
  console.log(`└─ Error:   ${styling(padLeft(status.error.toString(), maxChars, ' '), 'red')}`);
  console.log('');

  console.log(`Runtime: ${runtime} (mm:ss)`);
  console.log('');
};

export const outputProgress = (count: number): BarReturnType => {
  const bar = new SingleBar({
    format: `${styling('{bar}', 'red')} {percentage}% | {value}/{total} Files`,
    barCompleteChar: '\u2588',
    barIncompleteChar: '\u2591',
    hideCursor: true,
  });

  bar.start(count, 0);

  return {
    increment: () => bar.increment(),
    stop: () => {
      bar.stop();
      console.log('');
    },
  };
};

export const outputStatus = (countFiles: number, countTestFiles: number, cpuCount: number, cpuUsed: number) => {
  console.log('');
  console.log(`CPU detected: ${styling(cpuCount, 'red')} (${styling(cpuUsed, 'red')} are used)`);
  console.log(`Files found:  ${styling(countFiles, 'red')} (${styling(countTestFiles, 'red')} with tests)`);
  console.log('');
};

export const outputTable = (thresholds: ThresholdGroupType[], thresholdLimits: ThresholdType<number>) => {
  const thresholdRows = createTableData(thresholds, thresholdLimits);

  console.log(table(thresholdRows, tableConfig));
};
