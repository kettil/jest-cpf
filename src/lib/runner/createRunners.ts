import { createFolder } from '../files/fileSystems';
import { JestConfigType, RunnerType } from '../types';
import createRunner from './createRunner';

const createRunners = async (config: JestConfigType, cpuUsed: number) => {
  const mainFolder = await createFolder(config.rootDir, config.coverageDirectory);
  const promises: Array<Promise<RunnerType>> = [];

  for (let i = 1; i <= cpuUsed; i++) {
    promises.push(createRunner(i, mainFolder, config));
  }

  const runners = await Promise.all(promises);

  return runners;
};

export default createRunners;
