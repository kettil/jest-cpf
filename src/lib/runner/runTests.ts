import { getJestFilename } from '../files/jest';
import { delay } from '../helper';
import { outputProgress } from '../outputs';
import { RunnerType, JestConfigType, ThresholdGroupType, RunnerDataType } from '../types';
import runnerEvaluation from './runnerEvaluation';
import runnerInitialization from './runnerInitialization';

const runTests = async (
  testFiles: string[],
  runners: RunnerType[],
  argvArray: string[],
  configFile: string,
  config: JestConfigType,
) => {
  let isPending = false;

  const bar = outputProgress(testFiles.length);
  const command = await getJestFilename(config.rootDir);
  const thresholds: ThresholdGroupType[] = [];

  const data: RunnerDataType = {
    testFiles,
    argvArray,
    configFile,
    config,
    command,
    thresholds,
    barIncrement: bar.increment,
  };

  do {
    isPending = false;

    for (const runner of runners) {
      const isPendingEvaluation = await runnerEvaluation(runner, data);
      const isPendingInitialization = runnerInitialization(runner, data);

      isPending = isPending || isPendingEvaluation || isPendingInitialization;
    }

    if (isPending) {
      await delay(250);
    }
  } while (testFiles.length > 0 || isPending);

  bar.stop();

  return thresholds;
};

export default runTests;
