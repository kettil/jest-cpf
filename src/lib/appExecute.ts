import cleanUpFiles from './files/cleanUpFiles';
import getFileLists from './files/fileLists';
import { getJestConfig } from './files/jest';
import { cleanFileFromRootPath } from './helpers/string';
import normalizeConfig from './normalize';
import { outputStatus } from './outputs';
import createRunners from './runner/createRunners';
import runTests from './runner/runTests';
import { thresholdDefaultLimits } from './threshold/settings';
import { RunnerType, ArgvObjectType } from './types';

export const getFileListsCallback = (rootDirectory?: string) => {
  if (typeof rootDirectory === 'string') {
    return (path: string) => cleanFileFromRootPath(path, rootDirectory);
  }

  return () => '';
};

const appExecute = async (
  argvArray: string[],
  argvObject: ArgvObjectType,
  cwdDirectory: string,
  cpuCount: number,
  cpuUsed: number,
) => {
  if (typeof argvObject.config !== 'string') {
    throw new TypeError('The Argument "--config" is missing');
  }

  const jestConfig = await getJestConfig(cwdDirectory, argvObject.config);
  const { thresholdLimits, ...config } = normalizeConfig(jestConfig, cwdDirectory);

  const [files, testFiles, runners] = await Promise.all<string[], string[], RunnerType[]>([
    getFileLists([config.rootDir], config.collectCoverageFrom, getFileListsCallback()),
    getFileLists(config.roots, config.testMatch, getFileListsCallback(config.rootDir)),
    createRunners(config, cpuUsed),
  ]);

  const cleanedUpFiles = cleanUpFiles(files, testFiles, config);

  outputStatus(cleanedUpFiles.length, testFiles.length, cpuCount, cpuUsed);

  const thresholdData = await runTests(testFiles, runners, argvArray, argvObject.config, config);

  return { cleanedUpFiles, thresholdData, thresholdLimits: { ...thresholdDefaultLimits, ...thresholdLimits } };
};

export default appExecute;
