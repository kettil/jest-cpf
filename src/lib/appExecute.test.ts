import appExecute, { getFileListsCallback } from './appExecute';
import getFileLists from './files/fileLists';
import { getJestConfig } from './files/jest';
import { outputStatus } from './outputs';
import createRunners from './runner/createRunners';
import runTests from './runner/runTests';

jest.mock('./files/jest');
jest.mock('./files/fileLists');
jest.mock('./runner/createRunners');
jest.mock('./outputs');
jest.mock('./runner/runTests');

const getJestConfigMock = getJestConfig as jest.Mock;
const createRunnersMock = createRunners as jest.Mock;
const getFileListsMock = getFileLists as jest.Mock;
const outputStatusMock = outputStatus as jest.Mock;
const runTestsMock = runTests as jest.Mock;

describe('getFileListsCallback()', () => {
  test('it should work with rootDir', () => {
    const callback = getFileListsCallback('/path/to/project');

    expect(typeof callback).toBe('function');

    const result = callback('/path/to/project/src/index.ts');

    expect(result).toBe('src/index.ts');
  });

  test('it should work without rootDir', () => {
    const callback = getFileListsCallback();

    expect(typeof callback).toBe('function');

    const result = callback('/path/to/project/src/index.ts');

    expect(result).toBe('');
  });
});

describe('appExecute()', () => {
  test('it should work', async () => {
    getJestConfigMock.mockResolvedValueOnce({
      testMatch: ['**/*.test.{js,jsx,ts,tsx}'],
      collectCoverageFrom: ['src/**/*.{ts,tsx,js,jsx}'],
    });

    getFileListsMock.mockResolvedValueOnce(['/path/to/project/src/index.ts', '/path/to/project/src/app.ts']);
    getFileListsMock.mockResolvedValueOnce(['/path/to/project/src/index.test.ts']);
    createRunnersMock.mockResolvedValueOnce([{ isRunner: true }]);
    runTestsMock.mockResolvedValueOnce([
      {
        file: '/path/to/project/src/index.ts',
        threshold: {
          branches: { percent: 90, uncovered: 3 },
          statements: { percent: 90, uncovered: 3 },
          lines: { percent: 90, uncovered: 3 },
          functions: { percent: 90, uncovered: 3 },
        },
      },
    ]);

    const argvArray = ['--config', '/path/to/project/config.json'];
    const argvObject = { config: '/path/to/project/config.json' };

    const promise = appExecute(argvArray, argvObject, '/path/to/project', 5, 3);

    await expect(promise).resolves.toEqual({
      thresholdLimits: { branches: 95, functions: 95, lines: 95, statements: 95 },
      cleanedUpFiles: ['/path/to/project/src/index.ts', '/path/to/project/src/app.ts'],
      thresholdData: [
        {
          file: '/path/to/project/src/index.ts',
          threshold: {
            branches: { percent: 90, uncovered: 3 },
            functions: { percent: 90, uncovered: 3 },
            lines: { percent: 90, uncovered: 3 },
            statements: { percent: 90, uncovered: 3 },
          },
        },
      ],
    });

    expect(getJestConfigMock).toHaveBeenCalledTimes(1);
    expect(getFileListsMock).toHaveBeenCalledTimes(2);
    expect(createRunnersMock).toHaveBeenCalledTimes(1);
    expect(runTestsMock).toHaveBeenCalledTimes(1);

    expect(outputStatusMock).toHaveBeenCalledTimes(1);
    expect(outputStatusMock).toMatchSnapshot('outputStatus');
  });

  test('it should throw an error', async () => {
    const promise = appExecute([], {}, '/path/to/project', 5, 3);

    await expect(promise).rejects.toThrow('The Argument "--config" is missing');
  });
});
