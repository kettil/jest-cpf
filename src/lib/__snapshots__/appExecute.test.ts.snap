// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appExecute() it should work: outputStatus 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      2,
      1,
      5,
      3,
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
